File: chap02_homeostasis/stochastic_MM.py
# -------------------------------------------------------------------

#  Python code for a
#  Gillespie simulation of the MM process
#  reaction 1  S + E -> C at rate k1
#  reaction 2  C -> S + E at rate km1
#  reaction 3  C -> S + E at rate k2

#  For Chapter 2, Section 2.9.3 of
#  Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.

#  Written by James Keener and James Sneyd

# -------------------------------------------------------------------

import numpy as np
import matplotlib.pyplot as plt

# Parameters
N = 100   # Initial substrate number
E = 2     # Enzyme numbers
K = 2000  # Number of trials
k1 = 1
km1 = 1   # Without loss of generality
k2 = 0.5  # Without loss of generality

# Set the rate constants for three reactions
r = np.array([k1, km1, k2])

# Specify the change matrix
# Ch is the change matrix, a 3 by 2 matrix; three reactions, two state variables
Ch = np.array([[-1, 1], [1, -1], [0, -1]])

# Initialize the state space
s = np.ones(K) * N  # Start with N substrate molecules
c = np.zeros(K)     # Start with zero complex molecules
S = np.zeros((K, K))  # This will keep track of the trajectories
C = np.zeros((K, K))
T = np.zeros((K, K))  # This will track the transition times
j = 1  # Number of reaction steps

ndx = np.where(s + c > 0)[0]  # Update only the s's that are not zero
Nt = len(ndx)  # =K at first
Kt = 5000  # Max number of time steps to prevent very long runs without termination

while Nt > 0 and j < Kt:
    j = j + 1
    e = E - c
    h = np.vstack([r[0] * s * e, r[1] * c, r[2] * c]).T  # Reaction rates
    hc = np.cumsum(h, axis=1)  # The cumulative sum of h
    H = np.sum(h, axis=1)

    rn = np.random.rand(Nt, 2)  # Random numbers for the number of not yet terminated trajectories

    T[:, j] = T[:, j - 1]  # Add the current time to T

    for k in range(Nt):
        T[ndx[k], j] = -np.log(rn[k, 0]) / H[ndx[k]] + T[ndx[k], j - 1]  # Time of next reaction
        rk = np.where(rn[k, 1] <= hc[ndx[k], :] / H[ndx[k]])[0][0]  # Determine which reaction occurs
        s[ndx[k]] = s[ndx[k]] + Ch[rk, 0]  # Update s
        c[ndx[k]] = c[ndx[k]] + Ch[rk, 1]  # Update c

    # Save the values of the trajectories
    S[:, j] = s
    C[:, j] = c
    ndx = np.where(s + c > 0)[0]  # Check which trajectories are not extinct yet

    if len(ndx) == 0:
        Nt = 0
    else:
        Nt = len(ndx)  # Nt is the number of trajectories not yet extinct

# Use the first trial to plot sample trajectories
k=0

# Phase portrait of some trajectories
plt.figure(1)
plt.plot(T[k, 2:-1], S[k, 2:-1], T[k, 2:-1], C[k, 2:-1])
plt.xlabel('T', fontsize=20)
plt.legend(['S', 'C'], fontsize=18)
plt.title('Sample Trajectory', fontsize=20)

plt.figure(2)
plt.plot(S[k, 2:-1], C[k, 2:-1])
plt.xlabel('S', fontsize=20)
plt.ylabel('C', fontsize=18)
plt.title('Sample Trajectory', fontsize=20)

# Process the data
# Create the pdf for extinction times from the data
NN, TT = np.histogram(T[:, j], 50)  # Histogram of the extinction times with 50 boxes
tt = np.sort(TT)
dt = np.mean(TT[1:] - TT[:-1])  # Timestep increment on the histogram

# NN / (K * dt) is the approximate pdf for the extinction times from data
# It is normalized to have a total integral = 1

plt.figure(3)
NNne0 = np.where(NN > 0)[0]
plt.plot(tt[NNne0], NN[NNne0] / (K * dt), '*')
plt.xlabel('Completion time', fontsize=20)
plt.ylabel('Pdf of completion times', fontsize=20)
plt.show()

File: chap02_homeostasis/stochastic_MM.m
%-------------------------------------------------------------------

% Matlab code for a
% Gillespie simulation of the MM process
% reaction 1  S + E -> C at rate k1
% reaction 2  C -> S + E at rate km1
% reaction 3  C -> S + E at rate k2

% For Chapter 2, Section 2.9.3 of
% Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.

% Written by James Keener and James Sneyd

%-------------------------------------------------------------------

clear all

set(0,                           ...
'defaultaxesfontsize', 20,   ...
'defaultaxeslinewidth', 2.0, ...
'defaultlinelinewidth', 2.0);


N = 100; %initial substrate number
E = 2; %enzyme numbers

K = 2000;  % number of trials
k1 = 1;
km1 = 1; % without loss of generality
k2  = 0.5;    % without loss of generality

%set the rate constants for two reactions:
r(1) = k1;  % reaction 1  S + E -> C at rate k1 
r(2) = km1; % reaction 2  C -> S + E at rate km1
r(3) = k2;  % reaction 3  C -> P + E at rate k2

%Specify the change matrix
%Ch is the change matrix, a 3 by 2 matrix; three reactions, two state variables
Ch = [-1,1;1,-1;0,-1];  %What happens to s, c  when a reaction occurs:
%initialize the state space

s = N* ones(K,1); % start with N substrate molecules
c = zeros(K,1); %start zero complex molecules

S = s ; %This   keeps track of the trajectories
C = c;
T = zeros(K,1);  %This   tracks the transition times
j = 1; % number of reaction steps
rk = [];

ndx = find(s+c>0);  %update only the s's that are not zero
Nt = length(ndx); %=K at first
Kt = 5000; % max number of time steps to prevent very long runs without termination
while (Nt>0&j<Kt)
j = j+1;
e = E-c;
h(:,1) = r(1)*s.*e;  %reaction rate 1
h(:,2) = r(2)*c; % reaction rate 2
h(:,3) = r(3)*c;  % reaction rate 3 
hc = cumsum(h')'; % the cumulative sum of h
H = sum(h')';

rn = rand(Nt,2); %find a random number for the number of not yet terminated trajectories

T(:,j) = T(:,j-1);  % add the current time to T

for k = 1:Nt
 % update only those for which s+c >0
 
T(ndx(k),j) = - log(rn(k,1))/H(ndx(k)) +T(ndx(k),j-1); % time of next reaction
rk  =  find(rn(k,2) <=hc(ndx(k),:)/H(ndx(k)) ,1); % this determines which reaction occurs
s(ndx(k)) = s(ndx(k)) + Ch(rk,1); % update s 
c(ndx(k)) = c(ndx(k)) + Ch(rk,2); % update c

end 
% save the values of the  trajectories
S(:,j) = s ;
C(:,j) = c ;
ndx = find(s+c>0);  %check to see which trajectories are not extinct yet
if (isempty(ndx)==1)
    Nt = 0;
else
    Nt = length(ndx); %Nt is the number of trajectories not yet extinct

end
end

k = 1; % use the first several trials to plot   sample trajectories

% phase portrait of some trajectories
figure(1)
plot(T(1,:),S(1,:) ,T(1,:),C(1,:))
xlabel('T', 'fontsize', 20)
legend('S','C','fontsize',18)
title('Sample Trajectory','fontsize',20)

figure(2)
plot(S(1,:) , C(1,:))
xlabel('S', 'fontsize', 20)
ylabel( 'C','fontsize',18)
title('Sample Trajectory','fontsize',20)

% now process the data 
% create the pdf for extinction times from the data
[NN,TT]=hist(T(:,end),50);  % histogram of the extinction times with 50 boxes
tt = sort(TT);
dt = mean(TT(2:end)-TT(1:end-1));  %timestep increment on the histogram

% NN/(K*dt) is the approximate pdf for the extinction times from data
% it is normalized to have total integral = 1

figure(3)
NNne0=find(NN>0);
plot(TT(NNne0), NN(NNne0)/(K*dt),'*')
xlabel('Completion time','fontsize',20)
ylabel('Pdf of completion times','fontsize',20)
